### PR TITLE
fix: one K8s service per application

### PR DESCRIPTION
**Pull request by manadart**
The change where this became plural was added under https://github.com/juju/juju/pull/18360.

Unfortunately that change was made after approval contingent upon an agreed schema in which it was singular -- the diagram is available in one of my comments on that patch.

The fault is rectified here.

Migration tests verify roll-back upon detection of duplicates for K8s services. This is not anticipated as a likely occurrence, but would require a cleanse of the source prior to migration.

### CODE REVIEW

File: domain/schema/model/sql/0019-application.sql [hunk 1]
```diff
@@ -68,7 +68,7 @@ CREATE TABLE k8s_service (
 CREATE UNIQUE INDEX idx_k8s_service_provider
 ON k8s_service (provider_id);
 
-CREATE INDEX idx_k8s_service_application
+CREATE UNIQUE INDEX idx_k8s_service_application
```
Comment (SimonRichardson) [inline domain/schema/model/sql/0019-application.sql:71]: Ouch.

**Review by SimonRichardson — APPROVED** [review]

File: domain/schema/model/sql/0019-application.sql (hunk 1 above)
Comment (SimonRichardson) [reply domain/schema/model/sql/0019-application.sql:71]: This is why had weirdness in my original branch :face_exhaling:

Comment (gfouillet) [inline domain/schema/model/sql/0019-application.sql:71]: **question:** I believe that 4.0 might have duplicate there. I know that don't have 4.0 in production, is it acceptable ?

Moreover, why do we don't use a PATCH file here (to me PATCH files were mandatory except on main ?)

Comment (gfouillet) [inline domain/schema/model/sql/0019-application.sql]: **todo:** Refining lee-van-cleef's warning, verified: `checkSchemaHashesMatch` (`core/database/schema/query.go:113`) will abort any 4.1-beta model DB at its next `Ensure`/upgrade with `hash mismatch for version 19`, since `computeHashes` chains each patch's hash into the next. The blast radius is narrower than it sounds — in-place minor upgrades are not supported, so released 4.0 model DBs never run this check against the 4.1 chain; only 4.1-beta DBs are affected, which is accepted beta convention. However, the 4.0 backport cannot reuse this diff: it needs a new `.PATCH.sql` wired into `modelPostPatchFilesByVersion` (as 4.0 did for its 29 post-GA patches), and a bare `CREATE UNIQUE INDEX` would fail on live duplicate rows during the in-place upgrade — so the dedupe companion above must land first. Please flag the intended backport delivery before this merges.

File: domain/application/state/application.go [hunk 2]
```diff
@@ -1079,16 +1087,13 @@ AND    provider_id = $k8sService.provider_id`, k8sService{})
 			return errors.Errorf("cannot upsert cloud service for synthetic application %q", applicationName)
 		}
 
-		// First see if the cloud service for the app and provider id already exists.
-		// If so, it's a no-op.
 		serviceInfoToUpsert := k8sService{
 			ProviderID:      providerID,
 			ApplicationUUID: appDetails.UUID,
 		}
 		err = tx.Query(ctx, queryExistingStmt, serviceInfoToUpsert).Get(&serviceInfoToUpsert)
```
Comment (gfouillet) [inline domain/application/state/application.go:1094]: **nitpick:** The three-arm chain leads with the compound negation `err != nil && !errors.Is(err, sqlair.ErrNoRows)`, which forces the reader to deduce that the final `else if` implies a successful `Get`. The file's house pattern (e.g. `checkApplicationExists`) checks `ErrNoRows` first and real errors second — reordering to `ErrNoRows` → real error → provider-ID comparison would give three simple conditions and an explicit success arm. The logic is verified correct either way; this is purely readability.

File: domain/application/state/application.go [hunk 3]
```diff
@@ -1061,12 +1063,18 @@ func (st *State) UpsertK8sService(ctx context.Context, applicationName, provider
...
-FROM   k8s_service
-WHERE  application_uuid = $k8sService.application_uuid
-AND    provider_id = $k8sService.provider_id`, k8sService{})
+SELECT ks.* AS &k8sService.*
+FROM k8s_service AS ks
+WHERE ks.application_uuid = $k8sService.application_uuid`, k8sService{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+	updateProviderIDStmt, err := st.Prepare(`
```
Comment (gfouillet) [inline domain/application/state/application.go]: **nitpick:** This is the only aliased `UPDATE ... AS` in the codebase — sibling updates (`application.go:894`, `application.go:1820`, etc.) use the bare table name. The alias is defensible under the architecture rules' explicit-aliasing guidance, but note SQLite cannot alias-qualify `SET` columns, so the statement reads internally inconsistent (`SET provider_id` unqualified vs `WHERE ks.uuid` qualified). The unaliased form would match the ~25 sibling UPDATEs; either way it's cosmetic.

File: domain/application/state/application_test.go [hunk 4]
```diff
@@ -1176,34 +1176,102 @@ func (s *applicationStateSuite) TestUpsertK8sServiceExisting(c *tc.C) {
...
+			"SELECT ip.address_value FROM ip_address AS ip WHERE ip.net_node_uuid = ?", nodeUUID).Scan(&address)
+		c.Assert(err, tc.ErrorIsNil)
+		c.Check(address, tc.Equals, expected)
+	}
+	podInfo, err := s.state.GetUnitK8sPodInfo(c.Context(), "foo/0")
+	c.Assert(err, tc.ErrorIsNil)
+	c.Check(podInfo.Address, tc.Equals, "10.0.0.2")
+}
+
+func (s *applicationStateSuite) TestK8sServiceUniqueApplication(c *tc.C) {
```
Comment (gfouillet) [inline domain/application/state/application_test.go:1262]: **todo:** The new UPDATE error path (`updating cloud service provider ID: %w`, application.go:1109) is never exercised. A deterministic trigger is a provider-ID collision with another application's service via the pre-existing `idx_k8s_service_provider` UNIQUE index — upsert `foo` with `bar`'s provider ID and assert the full wrapped error chain plus that `foo`'s row still has its original provider ID and UUID (transaction rollback). This PR's UPDATE is what newly makes that constraint reachable from `UpsertK8sService`.

File: domain/application/state/application_test.go [hunk 5]
```diff
@@ -1176,34 +1176,102 @@ func (s *applicationStateSuite) TestUpsertK8sServiceExisting(c *tc.C) {
...
+	appUUID := s.createCAASApplication(c, "foo", life.Alive)
+	err := s.state.UpsertK8sService(c.Context(), "foo", "provider-id", nil)
+	c.Assert(err, tc.ErrorIsNil)
+
+	// Other writers, including migration, must also obey the invariant.
```
Comment (gfouillet) [inline domain/application/state/application_test.go:1267]: **todo:** The comment claims 'Other writers, including migration, must also obey the invariant', but no test exercises a duplicate-rows import payload — `TestCreateK8sServices` only feeds distinct applications and the migration coordinator test is gomocked. Once the dedupe companion lands (see the critical on `0019-application.sql:71`), please pin the chosen behaviour with an import test carrying two services for one application. (Cross-ref: lee-van-cleef's critical.)

File: domain/application/state/application.go [hunk 6]
```diff
@@ -1079,16 +1087,13 @@ AND    provider_id = $k8sService.provider_id`, k8sService{})
...
-		// If so, it's a no-op.
 		serviceInfoToUpsert := k8sService{
 			ProviderID:      providerID,
 			ApplicationUUID: appDetails.UUID,
 		}
 		err = tx.Query(ctx, queryExistingStmt, serviceInfoToUpsert).Get(&serviceInfoToUpsert)
 		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
-			return errors.Errorf(
-				"querying cloud service for application %q and provider id %q: %w", applicationName, providerID, err)
+			return errors.Errorf("querying cloud service: %w", err)
```
Comment (gfouillet) [inline domain/application/state/application.go]: **thought:** Counter-point to lee-van-cleef's suggestion 1 — the application name is not actually lost: the outer wrap at line 1122 (unchanged by this PR) adds `updating cloud service for application %q` to every error escaping the function, so the old inner message duplicated the identifier. The new generic message complies with the repo rule to add identifying context once at the highest useful layer; only the caller-supplied provider ID dropped out of the chain. If any context is worth restoring, it's the provider ID, not the application name.

File: domain/application/state/application.go [hunk 7]
```diff
@@ -1061,12 +1063,18 @@ func (st *State) UpsertK8sService(ctx context.Context, applicationName, provider
 		return errors.Capture(err)
 	}
 
-	// Query any existing records for application and provider id.
+	// An application has one cloud service, independent of its provider ID.
```
Comment (gfouillet) [inline domain/application/state/application.go]: **thought:** Counter-point to lee-van-cleef's suggestion 2 — a single `INSERT ... ON CONFLICT (application_uuid) DO UPDATE` isn't a drop-in replacement here. The insert path must create the `net_node` FK row before the service row (and must not create one on the update path — this PR's test asserts the node count is unchanged); the address upsert consumes the retained `serviceInfoToUpsert.UUID`/`NetNodeUUID`, which a bare ON CONFLICT doesn't surface without `RETURNING`; and a `provider_id` uniqueness collision fails identically either way. The SELECT-then-branch is the correct shape — the reorder nit on line 1094 is the only simplification worth taking.

**Review by gfouillet — APPROVED** [review]
I have one question, and few more test request. Looks good so far.

File: domain/schema/model/sql/0019-application.sql (hunk 1 above)
Comment (manadart) [reply domain/schema/model/sql/0019-application.sql:71]: Patches are mandatory if upgrading in place is required. So the the rule is _if it released with a version > beta_, so release candidates and greater.

Comment (manadart) [reply domain/schema/model/sql/0019-application.sql:71]: Added test ensuring roll-back on duplicate detection, and some text regarding this to the description.
